### PR TITLE
Add documentation extension point

### PR DESCRIPTION
**Pull request by VincentDondain**
`IErrorDocumentationProvider` has one method: `GetDocumentationLink` which gives the address of the documentation for a given error code within the error's description.

Related to https://github.com/xamarin/md-addins/pull/1080.

### CODE REVIEW

**Comment by VincentDondain** [comment]
@Therzok @mrward feel free to comment it's my first extension point (:

**Comment by VincentDondain** [comment]
@mrward According to what we discussed on slack I updated the `ErrorListPad` to also redirect to the documentation like my extension point is providing.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs [hunk 1]
```diff
@@ -235,15 +235,37 @@ public TaskListEntry (BuildError error, object owner)
...
 		{
 			if (!file.IsNullOrEmpty) {
 				if (System.IO.File.Exists (file)) {
 					var project = WorkspaceObject as Project;
 					IdeApp.Workbench.OpenDocument (file, project, Math.Max (1, line), Math.Max (1, column));
 				} else {
-					var pad = IdeApp.Workbench.GetPad<ErrorListPad> ()?.Content as ErrorListPad;
-					pad?.FocusOutputView ();
+					if (HasDocumentationLink ()) {
+						System.Diagnostics.Process.Start (DocumentationLink);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs]: `MonoDevelop.Ide.DesktopService.OpenUrl` here.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs [hunk 2]
```diff
@@ -539,21 +539,24 @@ void OnTaskCopied (object o, EventArgs args)
 		void OnShowReference (object o, EventArgs args)
 		{
 			string reference = null;
-			if (GetSelectedErrorReference (out reference)) {
-				Process.Start ("http://google.com/search?q=" + System.Web.HttpUtility.UrlEncode (reference));
-				return;
-			}
+			if (GetSelectedErrorReference (out reference) && reference != null)
+				Process.Start (reference);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs]: Possibly change this to `MonoDevelop.Ide.DesktopService.OpenUrl`

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs [hunk 3]
```diff
@@ -235,15 +235,37 @@ public TaskListEntry (BuildError error, object owner)
...
+		}
+
+		public bool HasDocumentationLink ()
+		{
+			bool documentationProvided = false;
+			var extensions = Mono.Addins.AddinManager.GetExtensionObjects<Extensions.IErrorDocumentationProvider> ("/MonoDevelop/Ide/IErrorDocumentationProvider", false);
+			for (var i = 0; !documentationProvided && i < extensions.Length; i++) {
+				var link = extensions [i].GetDocumentationLink (description);
+				if (!string.IsNullOrEmpty (link)) {
+					DocumentationLink = link;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs]: We need to settle on a standard, choose the first or the last. :) I'm not sure what the current one is, most of the code I've seen always chose the first extension which could handle a situation.

Comment (VincentDondain) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs]: Will be the first in this case, as soon as we get a non-null link we exit the loop (`documentationProvided = true`).

Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs]: I missed the condition in the for loop. The code is pretty verbose in this way, maybe we should just get rid of `documentationProvided` and do `return true` and `return false` to exit the loops? Code would look clearer this way.

Anyway, this is bikeshedding and preference only.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs (hunk 2 above)
Comment (VincentDondain) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs]: There isn't such thing as `MonoDevelop.Ide.DesktopService.OpenUrl`. Are you talking about Xwt `Desktop.OpenUrl`? `MacPlatform.OpenURL`?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs (hunk 3 above)
Comment (VincentDondain) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Tasks/TaskListEntry.cs]: I'm not a fan of multiple returns in general but I guess in this case it's okay indeed.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs (hunk 2 above)
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs]: I meant `DesktopService.ShowUrl`, sorry.

Comment (VincentDondain) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads/ErrorListPad.cs]: Ah yup, got it :P

**Comment by VincentDondain** [comment]
Comments addressed.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/IErrorDocumentationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+using System;
+namespace MonoDevelop.Ide.Extensions
+{
+	public interface IErrorDocumentationProvider
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/IErrorDocumentationProvider.cs]: Please use an abstract class instead of an interface.

**Comment by slluis** [comment]
The extension point should should allow specifying a regular expression in the extension node, which would be used to match the error code. It could also allow specifying an url with a placeholder for the error code. The goal is to be able to generate an error link from the extension data, and avoid loading the add-in assembly if not necessary.

**Pull request merged**